app/controller.go

```diff
@@ -98,53 +96,41 @@ func (h HTTPHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
-const headerContentType = "Content-Type"
-
 type HostForwarder struct {
-	AddressResolver HostAddressResolver
-	Client          *http.Client
+	URLResolver HostURLResolver
 }
 
 func (f *HostForwarder) Handler() HTTPHandler {
 	return func(w http.ResponseWriter, r *http.Request) error {
 		vars := mux.Vars(r)
 		zone := vars["zone"]
 		if zone == "" {
-			return NewBadRequestError("Invalid url, missing zone value", nil)
+			return fmt.Errorf("invalid url missing zone value: %q", r.URL.String())
 		}
 		host := vars["host"]
 		if host == "" {
-			return NewBadRequestError("Invalid url, missing host value", nil)
-		}
-		split := strings.SplitN(r.URL.String(), "hosts/"+host+"/", 2)
-		if len(split) != 2 {
-			return NewBadRequestError("Invalid url, missing host resource", nil)
-		}
-		addr, err := f.AddressResolver.GetHostAddr(zone, host)
-		if err != nil {
-			return err
-		}
-		url := fmt.Sprintf("%s/%s", addr, split[1])
-		req, err := http.NewRequest(r.Method, url, r.Body)
-		if err != nil {
-			return nil
-		}
-		resp, err := f.Client.Do(req)
-		if err != nil {
-			return err
+			return fmt.Errorf("invalid url missing host value: %q", r.URL.String())
 		}
-		defer resp.Body.Close()
-		b, err := ioutil.ReadAll(resp.Body)
+		hostURL, err := f.URLResolver.GetHostURL(zone, host)
 		if err != nil {
 			return err
 		}
-		w.Header().Set(headerContentType, resp.Header.Get(headerContentType))
-		w.WriteHeader(resp.StatusCode)
-		w.Write(b)
+		split := strings.SplitN(r.URL.Path, "hosts/"+host, 2)
```

Comment:
I think the `Director` function is the most natural place to do these URL transformations: It's purpose is to transform one URL into the other one after all

Reply:
acknowledged, done.

Reply:
I'm sorry I wasn't clear enough in my comment. My suggestion is to create a proxy object that could be re-used between requests. In the `Director` function it will extract variables from the request (with `mux.Var()`) and figure out the hostURL.

Reply:
I liked the idea of having a single proxy instance and managing the routing logic into the Director function, however didn't find it feasible due the Director function does not expect error in its logic, and fetching the host url could result in errors, hence I decided to create reverse proxy instances only when the request is valid, created a hostReverseProxy type though in case we want to cache instances later on.

Reply:
you could have an error returning logic from the `Director` logic by simply changing the url to an error page, but that's not worth it just to have a single proxy object